agent/dns.go

```diff
@@ -782,6 +782,11 @@ func (d *DNSServer) doDispatch(network string, remoteAddr net.Addr, req, resp *d
 
 	case "addr":
 		// <address>.addr.<suffixes>.<domain> - addr must be the second label, datacenter is optional
+
+		//check if the query type is A for IPv4 or AAAA for IPv6 or ANY
+		if (req.Question[0].Qtype != dns.TypeANY) && (req.Question[0].Qtype != dns.TypeA || len(queryParts[0])/2 != 4) && (req.Question[0].Qtype != dns.TypeAAAA || len(queryParts[0])/2 != 16) {
+			return invalid()
```

Comment:
`invalid()` is returning this:
```
RcodeNameError      = 3  // NXDomain  - Non-Existent Domain 
```

I wonder if that is the right response code. The issue reported said they exposed `NoError` (`RcodeSuccess`) with an empty body, but I wonder if `RcodeNotImplemented` would be appropriate.

Reply:
No it's not, I need to add logic to check which error to return:
- if the domain is not correct return invalid
- if the type is not implemented, we should return `not implemented`

Reply:
You should only return a `NXDomain` if there is no valid record at all for that query.  Thus if the `<address>` section would return a valid `A` or `AAAA` record, you should return as listed `NoError` (`RcodeSuccess`) instead.  If you return a NXDomain then resolvers will cache / return that even if a client also asks for a correct / existing record type.

Example if requesting `0b040cf9.addr.consul.` `AAAA` -> `NoError` & empty body.
`0b040cf9.addr.consul.` `A` -> `NoError` & `11.4.12.249` in answer.  If you respond to the `AAAA` request with `NXDomain` and it comes in first, the nameserver will also cache that for the default TTL time in the SOA and thus return `NXDomain` for the `A` request coming in a few ms later.  
If however something like `frog.addr.consul` was asked, you should return a `NXDomain`  (invalid) for every record type as there will never be a valid answer.